org.eclipse.jdt.launching/launching/org/eclipse/jdt/launching/JavaRuntime.java

```diff
@@ -3449,6 +3452,97 @@ public static VMStandin createVMFromDefinitionFile(File eeFile, String name, Str
+	 * @throws CoreException
+	 *             if an error occurs or a modular VM is query for packages of Java-8 or lower
+	 * @since 3.21
+	 */
+	public Collection<String> providedVMRuntimePackages(IVMInstall vm, IExecutionEnvironment environment) throws CoreException {
+		String release = environment != null ? environment.getProfileProperties().getProperty(JavaCore.COMPILER_COMPLIANCE) : null;
+		return providedVMRuntimePackages(vm, release);
+	}
+
+	/**
+	 * Returns the names of all packages provided by the given {@link IVMInstall vmInstall}, optionally restricted to the given
+	 * {@link IExecutionEnvironment execution-environment}.
+	 * <p>
+	 * The returned collection contains only distinct elements, sorted in alphabetical order.<br>
+	 * If the given VM-Install is {@link #isModularJava(IVMInstall) non modular}, the execution-environment is ignored.
+	 * </p>
+	 *
+	 * @param vm
+	 *            the vm-install to query
+	 * @param release
+	 *            the release to restrict the packages for, ignored for non-modular VMs
+	 * @return the distinct and alphabetically sorted collection of names of all packages in the VM
+	 * @throws CoreException
+	 *             if an error occurs or a modular VM is query for packages of Java-8 or lower
+	 * @since 3.21
+	 */
+	@SuppressWarnings("restriction")
+	public Collection<String> providedVMRuntimePackages(IVMInstall vm, String release) throws CoreException {
+		try {
+			if (vm instanceof AbstractVMInstall vmInstall) {
+				return vmInstall.systemPackagesCache.computeIfAbsent(release, r -> {
+					try {
+						return querySystemPackages(vmInstall, r);
+					} catch (CoreException e) {
+						throw new org.eclipse.jdt.internal.core.search.matching.MatchLocator.WrappedCoreException(e);
+					}
+				});
+			}
+		} catch (org.eclipse.jdt.internal.core.search.matching.MatchLocator.WrappedCoreException e) {
+			throw e.coreException;
+		}
+		return querySystemPackages(vm, release);
+	}
+
+	@SuppressWarnings("restriction")
+	private static final String JRT_PATH = "lib/" + org.eclipse.jdt.internal.compiler.util.JRTUtil.JRT_FS_JAR; //$NON-NLS-1$
+
+	@SuppressWarnings("restriction")
+	private Collection<String> querySystemPackages(IVMInstall vm, String release) throws CoreException {
+		if (!JavaRuntime.isModularJava(vm)) {
+			for (LibraryLocation libLocation : JavaRuntime.getLibraryLocations(vm)) {
+				IPath path = libLocation.getSystemLibraryPath();
+				if (path != null) {
+					try (ZipFile zip = new ZipFile(path.toFile())) {
+						return zip.stream().filter(e -> !e.isDirectory()).map(ZipEntry::getName).filter(n -> !n.startsWith("META-INF/")).map(n -> { //$NON-NLS-1$
+							int i = n.lastIndexOf('/');
+							return n.substring(0, i);
+						}).distinct().map(n -> n.replace('/', '.')).sorted().toList();
+					} catch (IOException e) {
+						throw new CoreException(Status.error("Failed to read packages in JVM library for " + vm + ", at " + path, e)); //$NON-NLS-1$//$NON-NLS-2$
+					}
+				}
+			}
```

Comment:
For non-modular, pre Java-8 VMs I assume JDT already has some code to read all the VM-jars (like rt.jar, resources.jar, ...), but I didn't manage to find that until now.
I'll continue searching, but if anybody has a hint, that's more than welcome.

Reply:
@jukzi any idea for this one?

Reply:
maybe something like org.eclipse.jdt.internal.core.builder.ClasspathJar.readJarContent(SimpleSet)

Reply:
Thanks for that hint.
I looked into it and tried `new ClasspathJar(zip, null, false)listPackages();`, but that returns a list that besides the packages that contain classes also contains all the possible prefixes of those packages with actual content.
For example for `java.util` this also returns just `java`.
But that is not wanted here and I don't see a way to filter them.

Furthermore ClasspathJar internally uses a PackageCache, which I'm a bit hesitate to use in the shade of https://github.com/eclipse-jdt/eclipse.jdt.core/issues/1526, especially since the method implemented here already uses its own cache.